File: voidhack_agent_firewall/langchain.py
import os
import sys
from typing import Any, Dict, List, Optional

# Add proxy directory to sys.path to access validation rules
sys.path.append(os.path.join(os.path.dirname(os.path.dirname(__file__)), "proxy"))

from app.detect.rules import check_tool_calls  # type: ignore
from app.schemas import ToolCall, Status  # type: ignore
from app.policy import load_policy  # type: ignore

# Attempt to import LangChain's callback base class
try:
    from langchain_core.callbacks import BaseCallbackHandler
except ImportError:
    # Fallback mock if langchain is not installed in the current environment
    class BaseCallbackHandler: # type: ignore
        pass

class PolicyViolationError(ValueError):
    """Raised when an agent action violates security policies."""
    pass

class FirewallCallbackHandler(BaseCallbackHandler):
    """LangChain callback plugin that halts agent runs when tool calls or arguments violate security policy."""
    def __init__(self, policy_path: str):
        self.policy = load_policy(policy_path)

    def on_tool_start(
        self,
        serialized: Dict[str, Any],
        input_str: str,
        *,
        run_id: Any = None,
        parent_run_id: Any = None,
        tags: Optional[List[str]] = None,
        metadata: Optional[Dict[str, Any]] = None,
        **kwargs: Any
    ) -> Any:
        """Intercepts tool execution right before it starts."""
        tool_name = serialized.get("name", "")
        
        # Check against denied tools
        if tool_name in self.policy.tool_denylist:
            raise PolicyViolationError(
                f"Security Block: Tool '{tool_name}' is restricted by security policy."
            )

        # Check against allowed tools & argument-level rules
        if self.policy.tool_allowlist and tool_name not in self.policy.tool_allowlist:
            raise PolicyViolationError(
                f"Security Block: Tool '{tool_name}' is not in the allowed tool list."
            )

        # Wrap in a ToolCall object to run argument checks
        # LangChain tools typically pass inputs as JSON string or keyword args
        tc = ToolCall(
            id=str(run_id or "lc-run"),
            type="function",
            function={"name": tool_name, "arguments": input_str}
        )

        findings, rules_check = check_tool_calls([tc], self.policy)
        blocked = [f for f in findings if f.status is Status.BLOCK]
        
        if blocked:
            reasons = "; ".join(blocked[0].reasons)
            raise PolicyViolationError(
                f"Security Block: Tool '{tool_name}' argument check failed. Reason: {reasons}"
            )

File: agent-defender-js/lib/langchain.js
/**
 * LangChain.js callback handler — intercepts tool execution and blocks
 * unauthorized actions before they run.
 *
 * Usage:
 *   import { FirewallCallbackHandler } from "voidhack-agent-defender/langchain";
 *
 *   const handler = new FirewallCallbackHandler({ policyPath: "policy.yaml" });
 *   const executor = AgentExecutor.fromAgentAndTools({
 *     agent, tools, callbacks: [handler]
 *   });
 */

import { loadPolicy } from "./policy.js";
import { checkToolCalls } from "./rules.js";

export class PolicyViolationError extends Error {
  constructor(message) {
    super(message);
    this.name = "PolicyViolationError";
  }
}

export class FirewallCallbackHandler {
  /**
   * @param {{ policyPath: string }} opts
   */
  constructor(opts) {
    if (!opts?.policyPath) {
      throw new Error("FirewallCallbackHandler requires opts.policyPath");
    }
    this.policy = loadPolicy(opts.policyPath);
    this.name = "AgentFirewall";
  }

  /**
   * Called by LangChain right before a tool executes.
   * @param {{ name?: string }} tool  — serialized tool metadata
   * @param {string} input             — tool input string
   * @param {string} [runId]
   */
  async handleToolStart(tool, input, runId) {
    const toolName = tool?.name || "";

    // Check denied tools
    if (this.policy.tool_denylist.includes(toolName)) {
      throw new PolicyViolationError(
        `Security Block: Tool '${toolName}' is restricted by security policy.`
      );
    }

    // Check allowed tools
    if (
      this.policy.tool_allowlist.length > 0 &&
      !this.policy.tool_allowlist.includes(toolName)
    ) {
      throw new PolicyViolationError(
        `Security Block: Tool '${toolName}' is not in the allowed tool list.`
      );
    }

    // Check argument-level rules
    const tc = {
      id: runId || "lc-run",
      function: { name: toolName, arguments: input || "" },
    };
    const { findings } = checkToolCalls([tc], this.policy);
    const blocked = findings.filter((f) => f.status === "block");

    if (blocked.length > 0) {
      const reasons = blocked[0].reasons.join("; ");
      throw new PolicyViolationError(
        `Security Block: Tool '${toolName}' argument check failed. Reason: ${reasons}`
      );
    }
  }
}
